### ISSUE TITLE
Type casting breaks the validation of element types in the array

### ISSUE THREAD

**Issue by KostyaTretyak** [microsoft/TypeScript#46965]
# Bug Report

### 🔎 Search Terms

- type casting

### 🕗 Version & Regression Information

4.6.0-dev.20211130

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.0-dev.20211130#code/MYewdgzgLgBAhgJwXAngLhgbxgBwSHAUwSnRjAFcBbAI2JgF8BtAXRgF4YmAoGLPAsVIYAjAwA0vfviIkyY+BHhgU4mAHp1MADwBaXTAAqKIjGBxoASzABzKQHIohAB5R7azTv0wAciBjE+AgQajQUsJZuSgDu+LbcLADc3EA)

### 💻 Code

```ts
const array: { property: number }[] = [
  {property: 1},
  {property: 1} as any, // <-- Type casting
  'text', // <-- No errors, but it's wrong
];
```

### 🙁 Actual behavior

No errors.

### 🙂 Expected behavior

Should show errors.

**Comment by MartinJohns** [comment]
Duplicate of #19541. Used search terms: `array any in:title`
